akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala

```diff
@@ -32,6 +32,23 @@ trait SourceQueue[T] {
 }
 
 /**
+ * This trait adds completion support to [[SourceQueue]].
+ */
+trait SourceQueueWithComplete[T] extends SourceQueue[T] {
+  /**
+   * Complete the stream normally. Use `watchCompletion` to be notified of this
+   * operation’s success.
+   */
+  def complete(): Unit
+
+  /**
+   * Complete the stream with a failure. Use `watchCompletion` to be notified of this
+   * operation’s success.
+   */
+  def fail(ex: Throwable): Unit
```

Comment:
Yes, that is correct. I guess we disagree on whether that should be fixed, though: there are many bugs that we don’t detect, beginning with `def !(msg: Any)(sender: ActorRef): Unit`.